location/__init__.py

```diff
@@ -0,0 +1,149 @@
+import os
+import sys
+import json
+import re
+import requests
+
+sys.path.append("../")
+from utils import db, ms
+
+from dotenv import load_dotenv
+
+load_dotenv()
+
+if os.environ.get("GOOGLE_MAPS_KEY"):
+    key = os.environ.get("GOOGLE_MAPS_KEY")
+    BASE_URL = f"https://maps.googleapis.com/maps/api/geocode/json?key={key}"
+
+
+def add_location_data(articles):
+    """Add location information to every article in a list
+
+    For every article in a list, extract the institution name.
+    Every institution name is looked-up in the Mongo database,
+    if there is already a name->location mapping for the institution,
+    skip it. If there isn't a mapping, make a Google Maps API call.
+
+    Once we have ensured every institution name has a corresponding
+    Mongo ID, iterate through articles, looking up each institution
+    name locaiton_mappings.
+
+    Return the list of articles
+    """
+    institutions = [a.get("institution") for a in articles]
+
+    # geocode all locations (this only peforms the
+    # operation if the location is not already in the
+    # database.)
+    new_locations = [l for l in get_new_locations(institutions) if l]
+
+    # push all these newly-fetched locations to the
+    # database
+    if len(new_locations) > 0:
+        batch_insert_locations(new_locations)
+
+    # get all mongo ids so we have a mapping from
+    # institution -> location_id, including the
+    # new ones we just pushed
+    location_mappings = get_mongo_ids()
+
+    # for every article, lookup related institution
+    # in our mapping, and add the object_id relating
+    # to its location data
+    for article in articles:
+        institution = article.get("institution", None)
+        if institution:
+            location_id = location_mappings.get(institution)
+            if location_id:
+                article["location_data"] = location_id
+        else:
+            article["location_data"] = None
+
+    return articles
+
+
+def get_new_locations(queries):
+    """Return a list of locations that are not in Mongo
+
+    For every institution in queries, check if the
+    institution exists in Mongo (with location_in_db).
+    
+    If it does exist, do nothing.
+
+    If it does not exist, geocode the query with Google
+    Maps and then add the returned location data to the
+    return list.
+    """
+    location_data = []
+    for query in queries:
+        # if location is already in the db, skip
+        if not location_in_db(query):
```

Comment:
these sorts of looped requests to mongo can be avoided with batchwise requests

e.g. save a boolean array at the beginning where you check the location statuses of all articles, then loop over that

Reply:
I see. This just now "clicked" for me when you said to batch requests